// starrail.js/dist/models/light_cone/LightCone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightConeBuilder = exports.LightCone = void 0;
const config_file_js_1 = require("config_file.js");
const LightConeData_1 = require("./LightConeData");
class LightCone {
    constructor(data, client) {
        this.client = client;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.lightConeData = new LightConeData_1.LightConeData(json.getAsNumber("tid"), this.client);
        this.level = json.getAsNumber("level");
        this.ascension = json.getAsNumberWithDefault(0, "promotion");
        this.superimposition = this.lightConeData.superimpositions[json.getAsNumber("rank") - 1];
        this.basicStats = this.lightConeData.getStatsByLevel(this.ascension, this.level);
        this.extraStats = this.lightConeData.getSuperimpositionStats(this.superimposition.level);
    }
    static builder() {
        return new LightConeBuilder();
    }
}
exports.LightCone = LightCone;
class LightConeBuilder {
    constructor() {
        this.data = {};
        this.lightConeData = null;
    }
    level(level) {
        this.data.level = level;
        return this;
    }
    ascension(ascension) {
        this.data.promotion = ascension;
        return this;
    }
    superimposition(superimposition) {
        this.data.rank = superimposition;
        return this;
    }
    lightCone(lightConeData) {
        this.data.tid = lightConeData.id;
        return this;
    }
    build(client) {
        return new LightCone(this.data, client);
    }
}
exports.LightConeBuilder = LightConeBuilder;
