// starrail.js/dist/models/light_cone/LightConeData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightConeData = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
const TextAssets_1 = require("../assets/TextAssets");
const Path_1 = require("../Path");
const LightConeExpType_1 = require("./LightConeExpType");
const ImageAssets_1 = require("../assets/ImageAssets");
const LightConeSuperimposition_1 = require("./LightConeSuperimposition");
const StatProperty_1 = require("../StatProperty");
class LightConeData {
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getExcelData("EquipmentConfig", this.id);
        if (!_data)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("LightCone", this.id);
        this._data = _data;
        const _itemData = client.cachedAssetsManager.getExcelData("ItemConfigEquipment", this.id);
        if (!_itemData)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("LightCone Item", this.id);
        this._itemData = _itemData;
        const json = new config_file_js_1.JsonReader(this._data);
        const itemJson = new config_file_js_1.JsonReader(this._itemData);
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("EquipmentName", "Hash"), this.client);
        this.description = new TextAssets_1.TextAssets(itemJson.getAsNumber("ItemBGDesc", "Hash"), this.client);
        this.itemDescription = new TextAssets_1.TextAssets(itemJson.getAsNumber("ItemDesc", "Hash"), this.client);
        this.stars = Number(json.getAsString("Rarity").slice(-1));
        this.path = new Path_1.Path(json.getAsStringWithDefault("Unknown", "AvatarBaseType"), this.client);
        this.maxAscension = json.getAsNumber("MaxPromotion");
        this.maxSuperimposition = json.getAsNumber("MaxRank");
        const skillId = json.getAsNumber("SkillID");
        const superimpositions = [];
        for (let i = 1; i <= this.maxSuperimposition; i++) {
            superimpositions.push(new LightConeSuperimposition_1.LightConeSuperimposition(skillId, i, this.client));
        }
        this.superimpositions = superimpositions;
        this.expType = new LightConeExpType_1.LightConeExpType(json.getAsNumber("ExpType"), this.client);
        this.expProvide = json.getAsNumber("ExpProvide");
        this.coinCost = json.getAsNumber("CoinCost");
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("ThumbnailPath"), this.client);
        this.cardImage = new ImageAssets_1.ImageAssets(json.getAsString("ImagePath"), this.client);
    }
    getStatsByLevel(ascension, level) {
        const ascensionData = this.client.cachedAssetsManager.getExcelData("EquipmentPromotionConfig", this.id, ascension);
        if (!ascensionData)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("LightCone Ascension", `${this.id}-${ascension}`);
        const ascensionJson = new config_file_js_1.JsonReader(ascensionData);
        return [
            new StatProperty_1.StatPropertyValue("BaseHP", ascensionJson.getAsNumber("BaseHP", "Value") + ascensionJson.getAsNumber("BaseHPAdd", "Value") * (level - 1), this.client),
            new StatProperty_1.StatPropertyValue("BaseAttack", ascensionJson.getAsNumber("BaseAttack", "Value") + ascensionJson.getAsNumber("BaseAttackAdd", "Value") * (level - 1), this.client),
            new StatProperty_1.StatPropertyValue("BaseDefence", ascensionJson.getAsNumber("BaseDefence", "Value") + ascensionJson.getAsNumber("BaseDefenceAdd", "Value") * (level - 1), this.client),
        ];
    }
    getSuperimpositionStats(superimposition) {
        return this.superimpositions[superimposition - 1].stats;
    }
}
exports.LightConeData = LightConeData;
