// starrail.js/dist/models/enka/StarRailCharacterBuild.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StarRailCharacterBuild = void 0;
const config_file_js_1 = require("config_file.js");
const enka_system_1 = require("enka-system");
const Character_1 = require("../character/Character");
class StarRailCharacterBuild extends enka_system_1.CharacterBuild {
    constructor(data, client, username, hash) {
        super();
        this._data = data;
        this.client = client;
        this.enkaUserInfo = { username: username, hash: hash };
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = json.getAsString("name");
        this.order = json.getAsNumber("order");
        this.isLive = json.getAsBoolean("live");
        this.isPublic = json.getAsBoolean("public");
        this.character = new Character_1.Character(json.getAsJsonObject("avatar_data"), client);
        this.imageUrl = json.getAsNullableString("image");
        this.hoyoType = json.getAsNumber("hoyo_type");
        this.url = `https://enka.network/u/${this.enkaUserInfo.username}/${this.enkaUserInfo.hash}/${this.character.characterData.id}/${this.id}`;
    }
}
exports.StarRailCharacterBuild = StarRailCharacterBuild;
