// starrail.js/dist/models/relic/RelicData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelicData = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
const RelicExpType_1 = require("./RelicExpType");
const RelicMainStatGroup_1 = require("./RelicMainStatGroup");
const RelicSubStatGroup_1 = require("./RelicSubStatGroup");
const RelicSet_1 = require("./RelicSet");
const TextAssets_1 = require("../assets/TextAssets");
const ImageAssets_1 = require("../assets/ImageAssets");
const RelicType_1 = require("./RelicType");
class RelicData {
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getExcelData("RelicConfig", this.id);
        if (!_data)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Relic", this.id);
        this._data = _data;
        const _itemData = client.cachedAssetsManager.getExcelData("ItemConfigRelic", this.id);
        if (!_itemData)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Relic Item", this.id);
        this._itemData = _itemData;
        const json = new config_file_js_1.JsonReader(this._data);
        const itemJson = new config_file_js_1.JsonReader(this._itemData);
        this.name = new TextAssets_1.TextAssets(itemJson.getAsNumber("ItemName", "Hash"), this.client);
        this.description = new TextAssets_1.TextAssets(itemJson.getAsNumber("ItemBGDesc", "Hash"), this.client);
        this.type = new RelicType_1.RelicType(json.getAsString("Type"), this.client);
        this.stars = Number(json.getAsString("Rarity").slice(-1));
        this.maxLevel = json.getAsNumber("MaxLevel");
        this.expType = new RelicExpType_1.RelicExpType(json.getAsNumber("ExpType"), this.client);
        this.mainStatGroup = new RelicMainStatGroup_1.RelicMainStatGroup(json.getAsNumber("MainAffixGroup"), this.client);
        this.subStatGroup = new RelicSubStatGroup_1.RelicSubStatGroup(json.getAsNumber("SubAffixGroup"), this.client);
        this.expProvide = json.getAsNumber("ExpProvide");
        this.coinCost = json.getAsNumber("CoinCost");
        this.icon = new ImageAssets_1.ImageAssets(itemJson.getAsString("ItemIconPath"), this.client);
        this.figureIcon = new ImageAssets_1.ImageAssets(itemJson.getAsString("ItemFigureIconPath"), this.client);
        this.set = new RelicSet_1.RelicSet(json.getAsNumber("SetID"), this.client);
    }
}
exports.RelicData = RelicData;
