// starrail.js/dist/models/relic/RelicSet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelicSet = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
const TextAssets_1 = require("../assets/TextAssets");
const ImageAssets_1 = require("../assets/ImageAssets");
const RelicSetBonus_1 = require("./RelicSetBonus");
const RelicData_1 = require("./RelicData");
class RelicSet {
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getExcelData("RelicSetConfig", this.id);
        if (!_data)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("RelicSet", this.id);
        this._data = _data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("SetName", "Hash"), this.client);
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("SetIconPath"), this.client);
        this.figureIcon = new ImageAssets_1.ImageAssets(json.getAsString("SetIconFigurePath"), this.client);
        this.setBonus = json.get("SetSkillList").mapArray((_, needCount) => new RelicSetBonus_1.RelicSetBonus(this.id, needCount.getAsNumber(), this.client));
    }
    static getActiveSetBonus(relics) {
        const relicSets = relics.map(a => (a instanceof RelicSet) ? a : (a instanceof RelicData_1.RelicData) ? a.set : a.relicData.set);
        const separated = (0, config_file_js_1.separateByValue)(relicSets, a => a.id.toString());
        const relicSetCounts = Object.values(separated).map(array => { return { count: array.length, set: array[0] }; });
        return relicSetCounts.map(obj => {
            return {
                set: obj.set,
                count: obj.count,
                activeBonus: obj.set.setBonus.filter(bonus => bonus.needCount <= obj.count),
            };
        }).sort((a, b) => b.count - a.count);
    }
}
exports.RelicSet = RelicSet;
