// starrail.js/dist/models/relic/RelicSubStatGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelicSubStatGroup = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../../errors/AssetsNotFoundError");
const StatProperty_1 = require("../StatProperty");
class RelicSubStatGroup {
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getExcelData("RelicSubAffixConfig")[this.id];
        if (!_data)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("RelicSubStatGroup", this.id);
        this._data = _data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.subStats = json.mapObject((_, v) => {
            return {
                id: v.getAsNumber("AffixID"),
                groupId: this.id,
                statProperty: new StatProperty_1.StatProperty(v.getAsString("Property"), this.client),
                baseValue: v.getAsNumber("BaseValue", "Value"),
                stepValue: v.getAsNumber("StepValue", "Value"),
            };
        });
    }
}
exports.RelicSubStatGroup = RelicSubStatGroup;
