// starrail.js/dist/models/CombatType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.combatTypeIconColors = exports.CombatType = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = require("../errors/AssetsNotFoundError");
const TextAssets_1 = require("./assets/TextAssets");
const ImageAssets_1 = require("./assets/ImageAssets");
class CombatType {
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getExcelData("DamageType", this.id);
        if (!_data)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("CombatType", this.id);
        this._data = _data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("DamageTypeName", "Hash"), this.client);
        this.description = new TextAssets_1.TextAssets(json.getAsNumber("DamageTypeIntro", "Hash"), this.client);
        this.color = parseInt(json.getAsString("Color").replace(/^#/, ""), 16);
        this.iconColor = exports.combatTypeIconColors[this.id];
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("DamageTypeIconPath"), this.client);
        // use MazeEnterBattleWeakIconPath instead of DamageTypeIconPath as the former is bigger
        this.bigIcon = new ImageAssets_1.ImageAssets(json.getAsString("MazeEnterBattleWeakIconPath"), this.client);
    }
}
exports.CombatType = CombatType;
exports.combatTypeIconColors = {
    Physical: 0xCFCFCF,
    Fire: 0xEE4639,
    Ice: 0x4CABDE,
    Thunder: 0xC65EE2,
    Wind: 0x5FCC97,
    Quantum: 0x6A65CC,
    Imaginary: 0xFAE762,
};
