// starrail.js/dist/models/Path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Path = exports.pathMap = void 0;
const config_file_js_1 = require("config_file.js");
const TextAssets_1 = require("./assets/TextAssets");
const AssetsNotFoundError_1 = require("../errors/AssetsNotFoundError");
const ImageAssets_1 = require("./assets/ImageAssets");
exports.pathMap = {
    Warrior: "Destruction",
    Rogue: "The Hunt",
    Mage: "Erudition",
    Shaman: "Harmony",
    Warlock: "Nihility",
    Knight: "Preservation",
    Priest: "Abundance",
    Unknown: "General",
};
class Path {
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getExcelData("AvatarBaseType", this.id);
        if (!_data)
            throw new AssetsNotFoundError_1.AssetsNotFoundError("Path", this.id);
        this._data = _data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.name = new TextAssets_1.TextAssets(json.getAsNumber("BaseTypeText", "Hash"), this.client);
        this.description = new TextAssets_1.TextAssets(json.getAsNumber("BaseTypeDesc", "Hash"), this.client);
        this.icon = new ImageAssets_1.ImageAssets(json.getAsString("BaseTypeIcon"), this.client);
        this.smallIcon = new ImageAssets_1.ImageAssets(json.getAsString("BaseTypeIconSmall"), this.client);
    }
}
exports.Path = Path;
